nikola/image_processing.py

```diff
@@ -81,69 +81,94 @@ def filter_exif(self, exif, whitelist):
 
-        size = w, h = im.size
-        if w > max_size or h > max_size:
-            size = max_size, max_size
+        max_sizes should be a list of sizes, and the image would be resized to fit in a
+        square of each size (preserving aspect ratio).
 
-            # Panoramas get larger thumbnails because they look *awful*
-            if bigger_panoramas and w > 2 * h:
-                size = min(w, max_size * 4), min(w, max_size * 4)
+        dst_paths is a list of the destination paths, and should be the same length as max_sizes.
 
-        try:
-            exif = piexif.load(im.info["exif"])
-        except KeyError:
-            exif = None
-        # Inside this if, we can manipulate exif as much as
-        # we want/need and it will be preserved if required
-        if exif is not None:
-            # Rotate according to EXIF
-            value = exif['0th'].get(piexif.ImageIFD.Orientation, 1)
-            if value in (3, 4):
-                im = im.transpose(Image.ROTATE_180)
-            elif value in (5, 6):
-                im = im.transpose(Image.ROTATE_270)
-            elif value in (7, 8):
-                im = im.transpose(Image.ROTATE_90)
-            if value in (2, 4, 5, 7):
-                im = im.transpose(Image.FLIP_LEFT_RIGHT)
-            exif['0th'][piexif.ImageIFD.Orientation] = 1
+        Backwards compatibility:
 
-        try:
-            icc_profile = im.info.get('icc_profile') if preserve_icc_profiles else None
-            im.thumbnail(size, Image.ANTIALIAS)
-            if exif is not None and preserve_exif_data:
-                # Put right size in EXIF data
-                w, h = im.size
-                if '0th' in exif:
-                    exif["0th"][piexif.ImageIFD.ImageWidth] = w
-                    exif["0th"][piexif.ImageIFD.ImageLength] = h
-                if 'Exif' in exif:
-                    exif["Exif"][piexif.ExifIFD.PixelXDimension] = w
-                    exif["Exif"][piexif.ExifIFD.PixelYDimension] = h
-                # Filter EXIF data as required
-                exif = self.filter_exif(exif, exif_whitelist)
-                im.save(dst, exif=piexif.dump(exif), icc_profile=icc_profile)
-            else:
-                im.save(dst, icc_profile=icc_profile)
-        except Exception as e:
-            self.logger.warning("Can't process {0}, using original "
-                                "image! ({1})".format(src, e))
-            utils.copy_file(src, dst)
+        * If max_sizes is None, it's set to [max_size]
+        * If dst_paths is None, it's set to [dst]
+        * Either max_size or max_sizes should be set
+        * Either dst or dst_paths should be set
+        """
+
+        if dst_paths is None:
+            dst_paths = [dst]
+        if max_sizes is None:
+            max_sizes = [max_size]
+        if len(max_sizes) != len(dst_paths):
+            raise ValueError(f'resize_image called with incompatible arguments: {dst_paths} / {max_sizes}')
```

Comment:
We still support 3.5, which does not have f-strings.

Reply:
Should we? Changing it anyway :-)

Reply:
3.5 is generally usable, and Ubuntu 16.04 is still supported.

Reply:
Oh, well. Changed.